doc/jjfs.1
.TH JJFS 1

.TH PWLD 1
.SH NAME
jjfs \- Joakim Jalap's file system
.SH SYNOPSIS
.B jjfs --entry \fIentry\fR [jjfs_opts] [fuse_opts] mountpoint
.BR
.SH DESCRIPTION
Jjfs is a fuse filesystem meant for a specific use case. It is meant to mount a
remote filesystem over sftp (a bit like sshfs(1)) but to minimize disk activity
on the remote system. Therefore a cache is kept on the local system and
consulted for stats. When reading of a file is requested this file is
transferred over sftp to a staging directory on the local machine and then read
from there. This is to minimize disk activity on the remote end (in the temporal
sense).
.SH OPTIONS
jjfs options:
.TP
.BR \-V, \-\-version
Print version and exit.
.TP
.BR \-h, \-\-help
Print help and exit.
.TP
.BR \-s\ \fIserver\fR, \-\-server\ \fIserver\fR, \-o\ server=\fIserver\fR
Server to connect to.
.TP
.BR \-p\ \fIport\fR, \-\-port\ \fIport\fR, \-o\ port=\fIport\fR
Port to connect to.
.TP
.BR \-u\ \fIuser\fR,\-\-user\ \fIuser\fR, \-o\ user=\fIuser\fR
User to connect as.
.TP
.BR \-o\ cache\-file=\fIcache\-file\fR
The file to store the cache in. Default is \fI~/.jjfs/<name-of-entry>-cache\fR.
.TP
.BR \-c\ \fIconfiguration file.\fR, \-\-conf-file \fIconfiguration file.\fR
.BR \-o\ conf\-file=\fIconfiguration file.\fR
File to read configuration from. Default is ~/.jjfs.conf. For a description of
the format, see section FILES.
.TP
.BR \-\-staging-dir\ \fIstaging directory\fR, \-o\ staging-dir=\fIstaging directory\fR
Directory where fetched files are stored. This directory must have sufficient
space available to store any file opened.
.SH FILES
.TP
.BR Configuration\ file.
jjfs will attempt to read its configuration parameters from a configuration
file. If no \-c argument is given, the default location is ~/.jjfs.conf. This
file consists of a few global options and then \fIentries\fR.
.TP
.BR Cache\ file.
This is were jjfs stores its cache. There will be one cache file per mounted
file system. If no path for it is given on the command line or in the config
file, it will default to ~/.jjfs/<entry>-cache, where <entry> is the name of the
entry in the config file.
.TP
.BR Staging\ directory.
This is where fetched files will be stored while they are open. Files are
deleted when closed, so this directory should be empty when no files are open.
.SH BUGS
Tis program surely leaks memory, but its memory use should be modest enough that
it doesn't really matter.
.SH EXAMPLE
.TP
To use the values set for entry 'my-server' in ~/.jjfs.conf and mount on ~/movies
.sp 1
jjfs -e my-server ~/movies
.sp
.TP
Same as above but on a different port.
.sp 1
jjfs --port=50000 -e my-server ~/movies
.sp
.TP
Use the general settings for 'my-server', but mount a different directory at a
different location. This is not good. It's better to make a new entry in the
configuration file.
.sp 1
jjfs -e my-server --remote-dir /home/joakim/music --cache-file ~/..jjfs/music-cache ~/music
.sp
.SH AUTHOR
This man page was written by Joakim Jalap (joakim.jalap@fastmail.com).









src/jjfs.c
/*
 * Copyright (C) 2016  Joakim Jalap
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include "config.h"
#include "jjfs_conf.h"
#include "jjfs_misc.h"
#include "jjfs_sftp.h"
#include "jjfs_cache.h"
#include "jjfs_fuse.h"

struct fuse_operations jjfs_oper = {
  .getattr = jjfs_getattr,
  /* .fgetattr = jjfs_fgetattr, */
  .read = jjfs_read,
  /* .read_buf = jjfs_read_buf, */
  .open = jjfs_open,
  .release = jjfs_release,
  .opendir = jjfs_opendir,
  .readdir = jjfs_readdir,
  .releasedir = jjfs_releasedir,
  .init = jjfs_init,
  .destroy = jjfs_destroy,
  .access = jjfs_access,
};

enum {
  KEY_HELP,
  KEY_VERSION,
};

#define JJFS_OPT(t, p, v)                \
  {t, offsetof(struct jjfs_args, p), v}

static struct fuse_opt jjfs_opts[] = {
  JJFS_OPT("--entry %s", entry, 0),
  JJFS_OPT("-e %s", entry, 0),  
  JJFS_OPT("--server %s", server, 0),  
  JJFS_OPT("server=%s", server, 0),  
  JJFS_OPT("--user %s", user, 0),
  JJFS_OPT("user=%s", user, 0),  
  JJFS_OPT("-c %s", conf_file, 0),
  JJFS_OPT("--conf-file %s", conf_file, 0),
  JJFS_OPT("conf-file=%s", conf_file, 0),    
  JJFS_OPT("--port %i", port, 0),
  JJFS_OPT("-p %i", port, 0),  
  JJFS_OPT("cache-file=%s", cache_file, 0),  
  JJFS_OPT("port=%i", port, 0),
  JJFS_OPT("remote-dir=%s", remote_dir, 0),
  JJFS_OPT("--remote-dir %s", remote_dir, 0),  
  JJFS_OPT("staging-dir=%s", staging_dir, 0),
  JJFS_OPT("sshconfig=%s", sshconfig, 0),
  JJFS_OPT("prefetch-bytes=%i", prefetch_bytes, 0),  
  JJFS_OPT("--rebuild", rebuild, 1),
  JJFS_OPT("-r", rebuild, 1),  
  FUSE_OPT_KEY("-V", KEY_VERSION),
  FUSE_OPT_KEY("--version", KEY_VERSION),
  FUSE_OPT_KEY("-h", KEY_HELP),
  FUSE_OPT_KEY("--help", KEY_HELP),
  FUSE_OPT_END
};

static void echo_usage(const char *argv0) {
  printf("Usage: %s {mountpoint | {-r | --rebuild}} [options]\n"
         "\n"
         "general options:\n"
         "    -o opt [,opt ...]  mount options\n"
         "    -V | --version     print version\n"
         "    -h | --help        print help\n"
         "\n"
         "jjfs options:\n"
         "    --server | -s      server\n"
         , argv0);
}

static void echo_version() {

  printf(PACKAGE ", version " PACKAGE_VERSION
         ", please send any bugreports to "
         PACKAGE_BUGREPORT ".\n");
}

/**
 * This functions only handles those options which were declared with
 * FUSE_OPT_KEY
 *
 */
static int jjfs_opt_proc(void* data, const char *arg, int key,
                         struct fuse_args *outargs) {
  switch(key) {
  case KEY_HELP:
    echo_usage(outargs->argv[0]);
    fuse_opt_add_arg(outargs, "-ho");
    fuse_main(outargs->argc, outargs->argv, &jjfs_oper, NULL);
    exit(EXIT_SUCCESS);
  case KEY_VERSION:
    echo_version();
    fuse_opt_add_arg(outargs, "--version");
    fuse_main(outargs->argc, outargs->argv, &jjfs_oper, NULL);
    exit(EXIT_SUCCESS);
  }
  return 1;
}


int main(int argc, char **argv) {

  if (argc < 2) JJFS_DIE("Too few arguments");

  struct fuse_args fargs = FUSE_ARGS_INIT(argc, argv);

  struct jjfs_args jjfsargs;

  memset(&jjfsargs, 0, sizeof(jjfsargs));

  fuse_opt_parse(&fargs, &jjfsargs, jjfs_opts, jjfs_opt_proc);

  jjfs_read_conf(&jjfsargs);

  if (jjfs_is_rebuild()) {
    jjfs_cache_rebuild();
    
    exit(EXIT_SUCCESS);
  }

  return fuse_main(fargs.argc, fargs.argv, &jjfs_oper, NULL);
}
